File: sdks/python/veridag/builder.py
"""
Ergonomic, deterministic transaction builder for the Veridag Python SDK.
"""

from .types import Transaction, SignedTransaction, ObjectRef, TransferValueOp, Address, ResourceBudget
from .codec import encode_transaction
from .crypto import Keypair, derive_object_id


class TxBuilder:
    """Fluent builder for constructing and signing transactions."""

    def __init__(self, keypair: Keypair):
        self._keypair = keypair
        self._chain_id = 1
        self._nonce = 0
        self._expiry_epoch = 0xFFFFFFFFFFFFFFFF

    def chain(self, chain_id: int) -> "TxBuilder":
        self._chain_id = chain_id
        return self

    def nonce(self, nonce: int) -> "TxBuilder":
        self._nonce = nonce
        return self

    def expiry(self, epoch: int) -> "TxBuilder":
        self._expiry_epoch = epoch
        return self

    def transfer(self, to: Address, amount: int) -> SignedTransaction:
        from_addr = self._keypair.address()
        from_id = derive_object_id(from_addr, 0)

        tx = Transaction(
            protocol_version=1,
            chain_id=self._chain_id,
            sender=from_addr,
            nonce=self._nonce,
            expiry_epoch=self._expiry_epoch,
            declared_reads=[],
            declared_writes=[],
            capabilities=[],
            operation=TransferValueOp(
                from_ref=ObjectRef(id=from_id, expected=self._nonce),
                to=to,
                amount=amount,
            ),
            resource_budget=ResourceBudget(),
            metadata=b"",
        )

        payload = encode_transaction(tx)
        signature = self._keypair.sign("VERIDAG_TX_V1", payload)
        return SignedTransaction(tx=tx, signature=signature)

File: sdks/typescript/src/builder.ts
/**
 * Ergonomic, deterministic transaction builder matching the Rust TxBuilder.
 */

import { encodeTransaction } from "./codec.ts";
import { Keypair, deriveObjectId } from "./crypto.ts";
import type { Address, SignedTransaction, Transaction } from "./types.ts";

export class TxBuilder {
  private keypair: Keypair;
  private chainIdVal: bigint = 1n;
  private nonceVal: bigint = 0n;
  private expiryEpochVal: bigint = 0xffffffffffffffffn;

  constructor(from: Keypair) {
    this.keypair = from;
  }

  public chain(chainId: bigint | number): this {
    this.chainIdVal = BigInt(chainId);
    return this;
  }

  public nonce(nonce: bigint | number): this {
    this.nonceVal = BigInt(nonce);
    return this;
  }

  public expiry(epoch: bigint | number): this {
    this.expiryEpochVal = BigInt(epoch);
    return this;
  }

  public transfer(to: Address, amount: bigint | number): SignedTransaction {
    const fromAddr = this.keypair.address();
    const fromId = deriveObjectId(fromAddr, 0n);

    const tx: Transaction = {
      protocolVersion: 1n,
      chainId: this.chainIdVal,
      sender: fromAddr,
      nonce: this.nonceVal,
      expiryEpoch: this.expiryEpochVal,
      declaredReads: [],
      declaredWrites: [],
      capabilities: [],
      operation: {
        type: "TransferValue",
        from: {
          id: fromId,
          expected: this.nonceVal,
        },
        to,
        amount: BigInt(amount),
      },
      resourceBudget: {
        computeUnits: 0n,
        memoryBytes: 0n,
      },
      metadata: new Uint8Array(0),
    };

    const payload = encodeTransaction(tx);
    const signature = this.keypair.sign("VERIDAG_TX_V1", payload);

    return { tx, signature };
  }
}
